From 8d1f287853bd4904c2c3ac79d9138b101848419c Mon Sep 17 00:00:00 2001
From: Gu Chao <gux.chao@intel.com>
Date: Tue, 7 Jul 2020 12:13:03 +0800
Subject: [PATCH] DRV_DPM_SW-319: DPM add API dp_get_dc_config()

- add dummy cbm_dp_get_dc_config() to cqm driver
---
 drivers/net/datapath/dpm/datapath_api.c       | 24 ++++++++++++++++++
 drivers/net/ethernet/lantiq/cqm/cbm_wrapper.c |  6 +++++
 include/net/datapath_api.h                    | 36 +++++++++++++++++++++++++++
 include/net/lantiq_cbm_api.h                  | 12 +++++++++
 4 files changed, 78 insertions(+)

diff --git a/drivers/net/datapath/dpm/datapath_api.c b/drivers/net/datapath/dpm/datapath_api.c
index 047a3e7106ce..fade353d97eb 100644
--- a/drivers/net/datapath/dpm/datapath_api.c
+++ b/drivers/net/datapath/dpm/datapath_api.c
@@ -2867,6 +2867,30 @@ int dp_set_net_dev_ops(struct net_device *dev, void *ops_cb, int ops_offset,
 }
 EXPORT_SYMBOL(dp_set_net_dev_ops);
 
+int dp_get_dc_config(struct dp_dc_res *res, int flag)
+{
+	struct cbm_dc_res r;
+
+	if (!res)
+		return DP_FAILURE;
+
+	r.cqm_inst = res->inst;
+	r.dp_port = res->dp_port;
+	r.res_id = res->res_id;
+	r.alloc_flags = get_dp_port_info(res->inst, res->dp_port)->alloc_flags;
+
+	if (cbm_dp_get_dc_config(&r, flag)) {
+		pr_err("%s: %s failed, inst=%d, dp_port=%d, res_id=%d\n, alloc_flags=0x%x",
+		       __func__, "cbm_dp_get_dc_config",
+		       r.cqm_inst, r.dp_port, r.res_id, r.alloc_flags);
+		return DP_FAILURE;
+	}
+	res->rx_res = r.rx_res;
+	res->tx_res = r.tx_res;
+	return DP_SUCCESS;
+}
+EXPORT_SYMBOL(dp_get_dc_config);
+
 static int __init dp_init(void)
 {
 	return dp_basic_proc();
diff --git a/drivers/net/ethernet/lantiq/cqm/cbm_wrapper.c b/drivers/net/ethernet/lantiq/cqm/cbm_wrapper.c
index beda453af767..ec8777d5316b 100644
--- a/drivers/net/ethernet/lantiq/cqm/cbm_wrapper.c
+++ b/drivers/net/ethernet/lantiq/cqm/cbm_wrapper.c
@@ -563,6 +563,12 @@ s32 pon_deq_cntr_get(int port, u32 *count)
 }
 EXPORT_SYMBOL(pon_deq_cntr_get);
 
+int cbm_dp_get_dc_config(struct cbm_dc_res *res, int flag)
+{
+	return CBM_FAILURE;
+}
+EXPORT_SYMBOL(cbm_dp_get_dc_config);
+
 void register_cbm(const struct cbm_ops *cbm_cb)
 {
 	g_cbm_ops = cbm_cb;
diff --git a/include/net/datapath_api.h b/include/net/datapath_api.h
index 8b58d24ec4a5..73fcaf3921f1 100644
--- a/include/net/datapath_api.h
+++ b/include/net/datapath_api.h
@@ -2294,4 +2294,40 @@ void dp_net_dev_get_ethtool_stats(struct net_device *dev,
  */
 int dp_set_net_dev_ops(struct net_device *dev, void *ops_cb, int ops_offset,
 		       u32 flag);
+
+struct dp_dc_rx_res {
+	int in_alloc_ring_size; /*!< [out] rxin alloc buffer ring
+				 *         burst size supported by SoC
+				 */
+	int out_enq_ring_size;  /*!< [out] rxout enqueue ring size
+				 *         supported by SoC
+				 */
+	int rx_bufs;            /*!< [out] maximum rx buffers can be supported
+				 *         for prefetching in RXIN
+				 *         based on res_id
+				 */
+};
+
+struct dp_dc_tx_res {
+	int in_deq_ring_size;   /*!< [out] txin ring burst size
+				 *         supported by SoC
+				 */
+	int out_free_ring_size; /*!< [out] txout free buffer ring burst size
+				 *         supported by SoC
+				 */
+	int tx_bufs;            /*!< [out] maximum Tx buffers can be supported
+				 *         in txin ring based on
+				 *         specified res_id
+				 */
+};
+
+struct dp_dc_res {
+	int inst;    /*! <in> dp instance */
+	int dp_port; /*! <in> dp port id/lpid/ep */
+	int res_id;  /*! <in> resource ID */
+	struct dp_dc_rx_res rx_res;  /*! <out> rx resource */
+	struct dp_dc_tx_res tx_res;  /*! <out> tx resource */
+};
+
+int dp_get_dc_config(struct dp_dc_res *res, int flag);
 #endif /*DATAPATH_API_H */
diff --git a/include/net/lantiq_cbm_api.h b/include/net/lantiq_cbm_api.h
index 3cb253858f88..50bdc546f14d 100644
--- a/include/net/lantiq_cbm_api.h
+++ b/include/net/lantiq_cbm_api.h
@@ -1191,6 +1191,18 @@ int32_t cbm_get_egress_port_info(uint32_t cbm_port, uint32_t *tx_ch,
 				 uint32_t *flags);
 int cbm_buff_resv_bytes_get(int cbm_inst, int size);
 s32 cbm_get_mtu_size(u32 *mtu_size);
+
+struct cbm_dc_res {
+	int cqm_inst;    /*! <in> cqm instance */
+	int dp_port;     /*! <in> dp port id/lpid/ep */
+	int res_id;      /*! <in> resource ID */
+	int alloc_flags; /*! <in> port alloc_flags */
+	struct dp_dc_rx_res rx_res; /*! <out> rx resource */
+	struct dp_dc_tx_res tx_res; /*! <out> tx resource */
+};
+
+int cbm_dp_get_dc_config(struct cbm_dc_res *res, int flag);
+
 /**
  * cbm_dma_port_config_t
  * cbm_cpu_port_config_t
